substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -468,130 +476,160 @@ where
+				},
+				Ok(BlockStatus::KnownBad) => {
 					info!(
-						"💔 New peer {} with unknown genesis hash {} ({}).",
-						peer_id, best_hash, best_number,
+						"💔 New peer {peer_id} with known bad best block {best_hash} ({best_number})."
 					);
-					return Err(BadPeer(peer_id, rep::GENESIS_MISMATCH))
-				}
+					Some(Err(BadPeer(peer_id, rep::BAD_BLOCK)))
+				},
+				Ok(BlockStatus::Unknown) => {
+					if best_number.is_zero() {
+						info!(
+							"💔 New peer {} with unknown genesis hash {} ({}).",
+							peer_id, best_hash, best_number,
+						);
+						return Some(Err(BadPeer(peer_id, rep::GENESIS_MISMATCH)))
+					}
 
-				// If there are more than `MAJOR_SYNC_BLOCKS` in the import queue then we have
-				// enough to do in the import queue that it's not worth kicking off
-				// an ancestor search, which is what we do in the next match case below.
-				if self.queue_blocks.len() > MAJOR_SYNC_BLOCKS.into() {
+					// If there are more than `MAJOR_SYNC_BLOCKS` in the import queue then we
+					// have enough to do in the import queue that it's not worth kicking off
+					// an ancestor search, which is what we do in the next match case below.
+					if self.queue_blocks.len() > MAJOR_SYNC_BLOCKS.into() {
+						debug!(
+							target: LOG_TARGET,
+							"New peer {} with unknown best hash {} ({}), assuming common block.",
+							peer_id,
+							self.best_queued_hash,
+							self.best_queued_number
+						);
+						// Replace `PeerSyncState::New` peer.
+						self.peers.insert(
+							peer_id,
+							PeerSync {
+								peer_id,
+								common_number: self.best_queued_number,
+								best_hash,
+								best_number,
+								state: PeerSyncState::Available,
+							},
+						);
+						return None
+					}
+
+					// If we are at genesis, just start downloading.
+					let (new_state, req) = if self.best_queued_number.is_zero() {
+						debug!(
+							target: LOG_TARGET,
+							"New peer {peer_id} with best hash {best_hash} ({best_number}).",
+						);
+
+						(Some(PeerSyncState::Available), None)
+					} else {
+						if self.peer_pool.lock().try_reserve_peer(&peer_id) {
+							let common_best = std::cmp::min(self.best_queued_number, best_number);
+
+							debug!(
+								target: LOG_TARGET,
+								"New peer {} with unknown best hash {} ({}), searching for common ancestor.",
+								peer_id,
+								best_hash,
+								best_number
+							);
+
+							(
+								Some(PeerSyncState::AncestorSearch {
+									current: common_best,
+									start: self.best_queued_number,
+									state: AncestorSearchState::ExponentialBackoff(One::one()),
+								}),
+								Some(ancestry_request::<B>(common_best)),
+							)
+						} else {
+							trace!(
+								target: LOG_TARGET,
+								"`ChainSync` is aware of a new peer {peer_id} with unknown best \
+								 hash {best_hash} ({best_number}), but can't start ancestry search \
+								 as the peer is reserved by another syncing strategy.",
+							);
```

Comment:
The peer might also not exist in `PeerPool`?

Reply:
That's possible only due to logic error in the code, as the peers are added to the strategies and `PeerPool` at the same time.